File: platforms/harmony/entry/src/main/ets/keyboard/CandidateWrapPolicy.ts
/**
 * Row allocation for the candidate flow, ported from
 * platforms/android/java/app/msime/client/CandidateWrapPolicy.java.
 *
 * Kept separate from the view so the decision is testable without laying anything out: given the
 * measured width of each candidate, it answers which row each one belongs to.
 */
export class CandidateWrapPolicy {
  static shouldWrap(occupiedWidth: number, childWidth: number, availableWidth: number,
                    spacing: number): boolean {
    return occupiedWidth > 0 && occupiedWidth + spacing + childWidth > availableWidth;
  }

  static rows(availableWidth: number, spacing: number, childWidths: number[]): number[] {
    if (availableWidth < 0 || spacing < 0) {
      throw new Error('Invalid candidate wrap dimensions');
    }
    const rows: number[] = new Array<number>(childWidths.length);
    let row: number = 0;
    let occupied: number = 0;
    for (let index: number = 0; index < childWidths.length; index++) {
      const width: number = childWidths[index];
      if (width < 0) {
        throw new Error('Invalid candidate width');
      }
      if (CandidateWrapPolicy.shouldWrap(occupied, width, availableWidth, spacing)) {
        row++;
        occupied = width;
      } else {
        occupied = occupied === 0 ? width : occupied + spacing + width;
      }
      rows[index] = row;
    }
    return rows;
  }
}

File: platforms/android/java/app/msime/client/candidate/CandidateWrapPolicy.java
package app.msime.client;

/** Pure row-allocation policy shared by the Android candidate flow layout and JVM tests. */
public final class CandidateWrapPolicy {
    private CandidateWrapPolicy() {}

    static boolean shouldWrap(int occupiedWidth, int childWidth, int availableWidth, int spacing) {
        return occupiedWidth > 0
            && (long) occupiedWidth + spacing + childWidth > availableWidth;
    }

    public static int[] rows(int availableWidth, int spacing, int[] childWidths) {
        if (availableWidth < 0 || spacing < 0 || childWidths == null)
            throw new IllegalArgumentException("Invalid candidate wrap dimensions");
        int[] rows = new int[childWidths.length];
        int row = 0;
        int occupied = 0;
        for (int index = 0; index < childWidths.length; index++) {
            int width = childWidths[index];
            if (width < 0) throw new IllegalArgumentException("Invalid candidate width");
            if (shouldWrap(occupied, width, availableWidth, spacing)) {
                row++;
                occupied = width;
            } else {
                occupied = occupied == 0 ? width : occupied + spacing + width;
            }
            rows[index] = row;
        }
        return rows;
    }
}
